core/node/crypto/client.go

```diff
@@ -2,140 +2,404 @@ package crypto
-	defer span.End()
+	if ic.tracer != nil {
+		var span trace.Span
+		ctx, span = ic.tracer.Start(ctx, "eth_chainId")
+		defer span.End()
+	}
 
 	return ic.Client.ChainID(ctx)
 }
 
 func (ic *otelEthClient) BlockNumber(ctx context.Context) (uint64, error) {
-	ctx, span := ic.tracer.Start(ctx, "eth_blockNumber")
-	defer span.End()
+	if ic.tracer != nil {
+		var span trace.Span
+		ctx, span = ic.tracer.Start(ctx, "eth_blockNumber")
+		defer span.End()
+	}
 
 	return ic.Client.BlockNumber(ctx)
 }
 
 func (ic *otelEthClient) SendTransaction(ctx context.Context, tx *types.Transaction) error {
-	ctx, span := ic.tracer.Start(ctx, "eth_sendRawTransaction")
-	defer span.End()
-
-	span.SetAttributes(attribute.String("tx_hash", tx.Hash().String()))
-	if len(tx.Data()) >= 4 {
-		span.SetAttributes(attribute.String("func_selector", hex.EncodeToString(tx.Data()[:4])))
+	if ic.tracer != nil {
+		var span trace.Span
+		ctx, span = ic.tracer.Start(ctx, "eth_sendRawTransaction")
+		defer span.End()
+
+		span.SetAttributes(attribute.String("tx_hash", tx.Hash().String()))
+		data := tx.Data()
+		methodName := getMethodName(&data)
+		span.SetAttributes(attribute.String("method_name", methodName))
 	}
 
 	return ic.Client.SendTransaction(ctx, tx)
 }
 
 func (ic *otelEthClient) HeaderByNumber(ctx context.Context, number *big.Int) (*types.Header, error) {
-	ctx, span := ic.tracer.Start(ctx, "eth_getHeaderByNumber")
-	defer span.End()
+	if ic.tracer != nil {
+		var span trace.Span
+		ctx, span = ic.tracer.Start(ctx, "eth_getHeaderByNumber")
+		defer span.End()
+	}
 
 	return ic.Client.HeaderByNumber(ctx, number)
 }
 
 func (ic *otelEthClient) BlockByNumber(ctx context.Context, number *big.Int) (*types.Block, error) {
-	ctx, span := ic.tracer.Start(ctx, "eth_getBlockByNumber")
-	defer span.End()
-
+	if ic.tracer != nil {
+		var span trace.Span
+		ctx, span = ic.tracer.Start(ctx, "eth_getBlockByNumber")
+		defer span.End()
+	}
 	return ic.Client.BlockByNumber(ctx, number)
 }
 
-func (ic *otelEthClient) CallContract(ctx context.Context, msg ethereum.CallMsg, blockNumber *big.Int) ([]byte, error) {
-	ctx, span := ic.tracer.Start(ctx, "eth_callContract")
-	defer span.End()
-
-	if len(msg.Data) >= 4 {
-		span.SetAttributes(attribute.String("func_selector", hex.EncodeToString(msg.Data[:4])))
+// extractCallErrorStatus extracts the revert reason from an error if it is a contract error
+// with a revert reason. Otherwise, it will return "revert" for contract data errors, or
+// "fail" for other types of errors.
+func extractCallErrorStatus(err error) string {
+	if de, ok := err.(rpc.DataError); ok {
+		hexStr := de.ErrorData().(string)
+		hexStr = strings.TrimPrefix(hexStr, "0x")
+		revert, e := hex.DecodeString(hexStr)
+		if e == nil {
+			reason, e := abi.UnpackRevert(revert)
+			if e == nil {
+				return reason
+			}
+		}
+		return "revert"
 	}
+	return "fail"
+}
 
-	return ic.Client.CallContract(ctx, msg, blockNumber)
+func getMethodName(data *[]byte) string {
```

Comment:
Why is data `*[]byte` and not just `[]byte`?

Reply:
To avoid copy by value of the array.

Reply:
this a slice, array would be copied, but go slice itself is a reference type, and doesn't copy underlying buffer 

(ironically enough this sometimes leads to the bugs when underlying buffer is modified and `slices.Clone` should be used, but this is not the case here)